sdk/nodejs/provider/experimental/analyzer.ts

```diff
@@ -522,6 +537,123 @@ export class Analyzer {
 
         return parts.join(" ");
     }
+
+    /**
+     * Gets the Pulumi resource type information for a resource reference.
+     * A strong assumption is that the referenced resource class is in a package installed to node_modules
+     * and contains a standard Pulumi-generated SDK compiled into JavaScript. To find the resource type token,
+     * the function will attempt to find the JavaScript module file that contains the resource class, and then
+     * extract the type from the __pulumiType property of the resource class. To find the package version,
+     * the function will attempt to read the package.json file in the root directory of the referenced package.
+     * @returns Object containing packageName, packageVersion, and pulumiType token
+     * @throws Error if the resource type cannot be determined with detailed context information
+     */
+    private getResourceType(
+        context: { component: string; property: string; inputOutput: InputOutput; typeName?: string },
+        type: typescript.Type,
+    ): {
+        packageName: string;
+        packageVersion: string;
+        pulumiType: string;
+    } {
+        const symbol = type.getSymbol();
+        if (!symbol) {
```

Comment:
If we don't find a symbol here it's probably worth checking if there's an `aliasSymbol`.  I've found in https://github.com/pulumi/pulumi/pull/18654 that with unions of `<type> | undefined` symbol is not set but `aliasSymbol` is.  

Might be worth adding a test for `pulumi.Input<ResourceReference | undefined>` to test it.

Reply:
Added a test which is green. Chatted offline and decided to keep this until we hit an error here.